### PR TITLE
Fix overflow bug that occasionally occurs when supplying 0 base

### PR DESCRIPTION
**Pull request by kevincheng96**
This PR fixes a math overflow bug that occasionally occurs when supplying 0 base. I also added a unit test to verify the fix, as well as another unit test to show a weird quirk that can occur when withdrawing 0 base. 

**Bug details:**
In `supplyBase`, we calculate `dstPrincipalNew` using `dstPrincipal` (old principal value before amount is supplied). The formula is:

`int104 dstPrincipalNew = principalValue(presentValue(dstPrincipal) + signed104(amount))`

When `amount=0`, the formula condenses to:

`dstPrincipalNew = principalValue(presentValue(dstPrincipal))`

Due to the fact that both `principalValue` and `presentValue` round down in favor of the protocol, we can actually end up with a value of `dstPrincipalNew < dstPrincipal`. 

This breaks our assumption in `repayAndSupplyAmount` that `newPrincipal >= oldPrincipal` MUST be true. In the old code, this would cause `supplyAmount` returned from `repayAndSupplyAmount` to be an extremely large number (`uint104(-1)` to be precise), which would later cause an overflow during an addition operation. The new code now explicitly checks this assumption and sets both `repayAmount` and `supplyAmount` to 0 if the assumption is violated. We apply a similar check in `withdrawAndBorrowAmount` as well.

### CODE REVIEW

File: test/withdraw-test.ts [hunk 1]
```diff
@@ -193,6 +193,46 @@ describe('withdrawTo', function () {
...
+    });
+    // Weird quirk of round down behavior where `withdrawAmount` is 1 even though
+    // `amount` is 0. So no base leaves Comet (which is expected)
+    expect(event(s0, 2)).to.be.deep.equal({
+      Transfer: {
+        from: alice.address,
+        to: ethers.constants.AddressZero,
+        amount: 1n,
+      }
+    });
```
Comment (kevincheng96) [inline test/withdraw-test.ts:233]: Just wanted to highlight this weird interaction that can occasionally occur when withdrawing 0 base.

File: contracts/Comet.sol [hunk 2]
```diff
@@ -592,6 +592,9 @@ contract Comet is CometMainInterface {
      * @dev Note: The assumption `newPrincipal >= oldPrincipal` MUST be true
```
Comment (jflatow) [inline contracts/Comet.sol:592]: I think we can remove this now, or move the comment below to a dev note?

File: contracts/Comet.sol [hunk 3]
```diff
@@ -606,6 +609,9 @@ contract Comet is CometMainInterface {
      * @dev Note: The assumption `oldPrincipal >= newPrincipal` MUST be true
```
Comment (jflatow) [inline contracts/Comet.sol:609]: Same as above

File: contracts/Comet.sol (hunk 2 above)
Comment (kevincheng96) [reply contracts/Comet.sol]: Done.

File: contracts/Comet.sol (hunk 3 above)
Comment (kevincheng96) [reply contracts/Comet.sol]: Done.

File: contracts/Comet.sol [hunk 4]
```diff
@@ -589,9 +589,11 @@ contract Comet is CometMainInterface {
 
     /**
      * @dev The change in principal broken into repay and supply amounts
-     * @dev Note: The assumption `newPrincipal >= oldPrincipal` MUST be true
      */
     function repayAndSupplyAmount(int104 oldPrincipal, int104 newPrincipal) internal pure returns (uint104, uint104) {
+        // If the new principal is less than the old principal, then no amount has been repaid or supplied.
```
Comment (jflatow) [inline contracts/Comet.sol:594]: unbelievable nit but are these the only full-sentence comments in the contract?

**Review by jflatow — APPROVED** [review]
💪 🥇 🏆 🎣

File: contracts/Comet.sol (hunk 4 above)
Comment (jflatow) [reply contracts/Comet.sol:594]: actually they aren't. I'll clean it up in a separate PR (if at all)

Comment (kevincheng96) [reply contracts/Comet.sol]: Sure thing, I removed the `.` here

**Review by scott-silver — COMMENTED** [review]
🙌

File: test/supply-test.ts [hunk 5]
```diff
@@ -193,6 +193,45 @@ describe('supplyTo', function () {
     });
   });
 
+  // This is an edge-case that can occur when a user supplies 0 base.
```
Comment (scott-silver) [inline test/supply-test.ts:196]: really good explanation

**Review by antonina-cherednichenko — APPROVED** [review]
great work
